bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -596,3 +606,37 @@ func (r *KubeadmConfigReconciler) reconcileTopLevelObjectSettings(cluster *clust
 		log.Info("Altering ClusterConfiguration", "KubernetesVersion", config.Spec.ClusterConfiguration.KubernetesVersion)
 	}
 }
+
+// storeBootstrapData creates a new secret with the data passed in as input,
+// sets the reference in the configuration status and ready to true.
+func (r *KubeadmConfigReconciler) storeBootstrapData(ctx context.Context, config *bootstrapv1.KubeadmConfig, data []byte) error {
+	secret := &corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      config.Name,
+			Namespace: config.Namespace,
+			Labels: map[string]string{
+				clusterv1.ClusterLabelName: config.ClusterName,
+			},
+			OwnerReferences: []v1.OwnerReference{
+				{
+					APIVersion: bootstrapv1.GroupVersion.String(),
+					Kind:       "KubeadmConfig",
+					Name:       config.Name,
+					UID:        config.UID,
+					Controller: pointer.BoolPtr(true),
+				},
+			},
+		},
+		Data: map[string][]byte{
+			"value": data,
+		},
+	}
+
+	if err := r.Client.Create(ctx, secret); err != nil {
```

Comment:
shall we ignore the create error that secret may already exist, since now the original plaintext data is not removed?

Reply:
The method is designed to be called only once, if there is an error that wants to re-create the bootstrap data, it's definitely a bug that we might have to look at

Reply:
can you please help me understand, in line 169 we don't reset  config.Status.BootstrapData to nil, does that mean that switch condition will always be true?

Reply:
You're right, good catch! Need to fix that before this gets in. 

I was more speaking in the general case, not the migration case, the secret should only ever be created once.

/hold

Reply:
Fixed, thanks!